src/server/streamableHttp.ts

```diff
@@ -0,0 +1,368 @@
+        jsonrpc: "2.0",
+        error: {
+          code: -32700,
+          message: "Parse error",
+          data: String(error)
+        },
+        id: null
+      }));
+      this.onerror?.(error as Error);
+    }
+  }
+
+  /**
+   * Handles DELETE requests to terminate sessions
+   */
+  private async handleDeleteRequest(req: IncomingMessage, res: ServerResponse): Promise<void> {
+    if (!this.validateSession(req, res)) {
+      return;
+    }
+    await this.close();
+    res.writeHead(200).end();
+  }
+
+  /**
+   * Validates session ID for non-initialization requests when session management is enabled
+   * Returns true if the session is valid, false otherwise
+   */
+  private validateSession(req: IncomingMessage, res: ServerResponse): boolean {
+    const sessionId = req.headers["mcp-session-id"];
+
+    if (!sessionId) {
+      // Non-initialization requests without a session ID should return 400 Bad Request
+      res.writeHead(400, this._customHeaders).end(JSON.stringify({
+        jsonrpc: "2.0",
+        error: {
+          code: -32000,
+          message: "Bad Request: Mcp-Session-Id header is required"
+        },
+        id: null
+      }));
+      return false;
+    } else if ((Array.isArray(sessionId) ? sessionId[0] : sessionId) !== this._sessionId) {
+      // Reject requests with invalid session ID with 404 Not Found
+      res.writeHead(404, this._customHeaders).end(JSON.stringify({
+        jsonrpc: "2.0",
+        error: {
+          code: -32001,
+          message: "Session not found"
+        },
+        id: null
+      }));
+      return false;
+    }
+
+    return true;
+  }
+
+
+  async close(): Promise<void> {
+    // Close all SSE connections
+    this._sseResponseMapping.forEach((response) => {
+      response.end();
+    });
+    this._sseResponseMapping.clear();
+    this.onclose?.();
+  }
+
+  async send(message: JSONRPCMessage, options?: { relatedRequestId?: RequestId }): Promise<void> {
+    const relatedRequestId = options?.relatedRequestId;
+    if (relatedRequestId === undefined) {
+      throw new Error("relatedRequestId is required for Streamable HTTP transport");
+    }
+
+    const sseResponse = this._sseResponseMapping.get(relatedRequestId);
+    if (!sseResponse) {
+      throw new Error(`No SSE connection established for request ID: ${String(relatedRequestId)}`);
+    }
+
+    // Send the message as an SSE event
+    sseResponse.write(
+      `event: message\ndata: ${JSON.stringify(message)}\n\n`,
+    );
+
+    // If this is a response message with the same ID as the request, we can check
+    // if we need to close the stream after sending the response
+    if ('result' in message || 'error' in message) {
+      if (message.id === relatedRequestId) {
+        // This is a response to the original request, we can close the stream
+        // after sending all related responses
+        this._sseResponseMapping.delete(relatedRequestId);
+        sseResponse.end();
```

Comment:
I think we first need to check whether this `sseResponse` is being used by _other_ request IDs in the map, since there might have been multiple requests batched together.